### DISCUSSION TITLE
Multibyte transfer between AI-deck and Crazyflie

### DISCUSSION THREAD

**Discussion by ChenMathilda** [bitcraze/discussions#92] [Q&A]
Hi，
I try to implement multibyte transfer between AI-deck and Crazyflie based on this code：https://github.com/pulp-platform/AI-deck-workshop/tree/main/Hands-on/Session%203.

I have achieved four bytes of data reception in Crazyflie side, but with each execution of`make clean all run platform=board` command on AI-deck side, the byte order will change.I need to execute `make clean all run platform=board` four times on the AI-deck side to get the correct byte sequence(Try 4 in pic).And on the fourth try, I get the `0 0 0 0` sequence first (the part circled in red in the picture), this is a strange phenomenon.

I don't know why this happens. I speculate that it is an AI-deck problem, and I hope can get answer here.

AI-deck side: send  `1008601 in uint32_t =  [0,15,99,217] in uint8_t`
![image](https://user-images.githubusercontent.com/61350967/173361487-47312d9a-ff71-4695-aef6-9fe774f9b02e.png)

Here is my CF `main.c` code:
```
//main.c
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#include "app.h"
#include "FreeRTOS.h"
#include "task.h"
#include "debug.h"
#include "uart_dma_pulp.h"
#include "log.h"

#define DEBUG_MODULE "HELLOWORLD"
#define BUFFERSIZE 4 //changed 1->4

uint8_t aideckRxBuffer[BUFFERSIZE];
volatile uint8_t dma_flag = 0;
uint32_t log_counter=0;

void appMain()
{
	DEBUG_PRINT("Application started! \n");
	USART_DMA_Start(115200, aideckRxBuffer, BUFFERSIZE);
	while(1) {
		// DEBUG_PRINT("app_Main-while\n");
		vTaskDelay(M2T(100));
		if (dma_flag == 1)
		{
			dma_flag = 0;  // clear the flag
			DEBUG_PRINT("\tCounter_d: %d %d %d %d \n", aideckRxBuffer[0], aideckRxBuffer[1],aideckRxBuffer[2],aideckRxBuffer[3]);
			log_counter = aideckRxBuffer[0];
			memset(aideckRxBuffer, 0, sizeof(uint8_t)*BUFFERSIZE);  // clear the dma buffer
		}
	}

}


void __attribute__((used)) DMA1_Stream1_IRQHandler(void)
{
	uint8_t RxTemp; //changed
	DMA_Cmd(USARTx_RX_DMA_STREAM,DISABLE); //changed
	RxTemp = USARTx->SR;  //changed
	RxTemp = USARTx->DR; //changed

 	DMA_ClearFlag(DMA1_Stream1, UART3_RX_DMA_ALL_FLAGS);
 	dma_flag = 1;
	DMA_Cmd(USARTx_RX_DMA_STREAM,ENABLE); //changed
}

LOG_GROUP_START(log_test)
LOG_ADD(LOG_UINT32, test_variable_x, &log_counter)
LOG_GROUP_STOP(log_test)
```
here is the AIdeck side main code`uart_send_counter.c` 
```
// uart_send_counter.c
/* PMSIS includes */
#include "pmsis.h"

/* Variables used. */
uint32_t to_send;

void test_uart_helloworld(void)
{
    printf("Entering main controller\n");
    printf("sizeof(uint32_t):%d\n",sizeof(uint32_t));

    uint32_t errors = 0;
    struct pi_device uart;
    struct pi_uart_conf conf;

    /* Init & open uart. */
    pi_uart_conf_init(&conf);
    conf.enable_tx = 1;
    conf.enable_rx = 0;
    conf.baudrate_bps = 115200;
    pi_open_from_conf(&uart, &conf);
    if (pi_uart_open(&uart))
    {
        printf("Uart open failed !\n");
        pmsis_exit(-1);
    }

    while(1)
    {
        to_send = 1008601; //send [0,15,99,217] in uint8_t
        pi_uart_write(&uart, &to_send, 4);
        pi_time_wait_us(500000);
    }
    pi_uart_close(&uart);

    pmsis_exit(errors);
}

/* Program Entry. */
int main(void)
{
    printf("\n\n\t *** PMSIS Uart HelloWorld ***\n\n");
    return pmsis_kickoff((void *) test_uart_helloworld);
}
```

**Comment by knmcguire** [comment]
Hi! This is pretty old code by now but let's debug it for you as I think this should still work. @vladniculescu made this part of the workshop, so perhaps he can give some insights too!

Some additional questions first before continuing to provide solutions

- Which version of gap_sdk are you using?
- Which version of the crazyflie-firmware are you on now?

**Reply by ChenMathilda** [reply]
> Hi! This is pretty old code by now but let's debug it for you as I think this should still work. @vladniculescu made this part of the workshop, so perhaps he can give some insights too!
> 
> Some additional questions first before continuing to provide solutions
> 
> * Which version of gap_sdk are you using?
> * Which version of the crazyflie-firmware are you on now?

Hi，
gap_sdk = 3.9.1
crazyflie-firmware = commit c1aed0d1cc64abc5e5477fa73647fa0c8fc2e63c

**Reply by knmcguire** [reply]
It is a bit difficult to see from how long time ago the commit is. Could you see which tag is the newest so I know from which release it is based?

**Reply by ChenMathilda** [reply]
> It is a bit difficult to see from how long time ago the commit is. Could you see which tag is the newest so I know from which release it is based?

Actually, I used the latest version cf-firmware.
![image](https://user-images.githubusercontent.com/61350967/174256101-b5874958-84dd-4175-b32b-6fd34f5a6f3b.png)

**Comment by knmcguire** [comment]
As this tutorial is rather outdated by now it is difficult for us to give support on it. What I rather recommend you to do is to follow the[ latest aideck tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) and see if you can get something to work with the uart_send_char: https://github.com/bitcraze/aideck-gap8-examples/tree/master/examples/other/test_functionalities/uart_send_char.

If your purpose is only to receive info from the aideck on the crazyflie, perhaps it would be good to read-up on the new CPX protocol: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/getting-started/cpx/ which is made for inter MCU communication.

**Reply by ChenMathilda** [reply]
> As this tutorial is rather outdated by now it is difficult for us to give support on it. What I rather recommend you to do is to follow the[ latest aideck tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) and see if you can get something to work with the uart_send_char: https://github.com/bitcraze/aideck-gap8-examples/tree/master/examples/other/test_functionalities/uart_send_char.
> 
> If your purpose is only to receive info from the aideck on the crazyflie, perhaps it would be good to read-up on the new CPX protocol: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/getting-started/cpx/ which is made for inter MCU communication.

Thanks for you answer, I'll try

**Reply by knmcguire** [reply]
thanks! If you have any questions please let us know! Then better to make a new discussions

**Reply by Yingxiu-Chang** [reply]
Hi, 
I tried to compile the uart_send_char as the following:
sudo docker run --rm -v ${PWD}:/module aideck-with-autotiler tools/build/make-example examples/other/test_functionalities/uart_send_char image

But it shows 
Collecting numpy==1.22.3
  Using cached numpy-1.22.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
ERROR: imageio 2.17.0 has requirement pillow>=8.3.2, but you'll have pillow 7.2.0 which is incompatible.
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.5
    Uninstalling numpy-1.19.5:
      Successfully uninstalled numpy-1.19.5
Successfully installed numpy-1.22.3
gapy --target=ai_deck --platform=board --work-dir=/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS      run --image --binary=/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test 
usage: gen_flash_image [-h]
                       [--target {ai_deck,gapoc_b,gapuino_v2_spi,gapoc_a_v2_spi,gapuino_v2,gap9_v2,gapuino,pulp,gapuino_spi,gapoc_a_v2,pulp_rnnext,gapoc_b_v2,gapoc_a,gapuino_v3,gapuino_v3_spi,gapoc_b_v2_spi,pulp_nn,gap9}]
                       [--py-target PY_TARGET]
                       [--platform {board,rtl,gvsoc,fpga,netlist,pgpin}] [-v]
                       [--work-dir WORK_DIR] [--config-ini INI_CONFIGS]
                       [--config-opt CONFIG_ITEMS] [--boot-loader BOOT_LOADER]
                       [--partition-table PARTITIONTABLE]
                       [--partition partition-name partition-image partition-name partition-image]
                       [--block-size BLOCKSIZE] [--flash-size FLASHSIZE]
                       [--flash-type {hyper,spi}]
                       [--flash-stimuli-format {s,l,m}]
                       [--flash-stimuli-file {s,l,m}] [--encryte ENCRYPT]
                       [--aes_key AES_KEY] [--aes_iv AES_IV] [-o OUTPUT]
gen_flash_image: error: argument --boot-loader: can't open '/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test': [Errno 2] No such file or directory: '/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test'
make: *** [/gap_sdk/utils/rules/pulp_rules.mk:243: image] Error 2


Would you please guide me how to compile it?
Best regards
Hugh

**Reply by knmcguire** [reply]
Hi!

Could you start a new discussion please?
